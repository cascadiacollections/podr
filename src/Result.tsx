import { h, FunctionComponent } from 'preact';

function formatDuration(duration: number): string {
  return new Date(1000 * duration).toISOString().substr(11, 8);
}

function formatPubDate(isoString: string): string {
  return new Date(isoString.replace(/-/g, "/")).toDateString();
}

function stripHtml(html: string): string {
  const doc: Document = new DOMParser().parseFromString(html, 'text/html');
  return doc.body.textContent || '';
}

interface IEnclosure {
  duration: number;
  link: string;
}

export interface IFeedItem {
  guid: string;
  title: string;
  description: string;
  pubDate: string;
  enclosure: IEnclosure;
}

interface IResultProps {
  result: IFeedItem;
  onClick: (feedItem: IFeedItem) => void;
}

/* tslint:disable:variable-name*/
export const Result: FunctionComponent<IResultProps> = (props: IResultProps) => {
  const { onClick, result } = props;

  return (
    <li
      class={'result list-group-item list-group-item-action'}
      onClick={() => onClick(result)}
      tabIndex={0}>
      <h2 class='title' dangerouslySetInnerHTML={{ __html: result.title }} />
      <strong class='pubDate'>{formatPubDate(result.pubDate)}</strong>
      <strong>&nbsp;&bull;&nbsp;</strong>
      <strong class='duration'>{formatDuration(result.enclosure.duration)}</strong>
      <p class='description' dangerouslySetInnerHTML={{ __html: result.description }} />
    </li>
  );
};
