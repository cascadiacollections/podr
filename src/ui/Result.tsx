import { h, FunctionComponent } from 'preact';

function formatDuration(duration: number): string {
  return new Date(1000 * duration).toISOString().substr(11, 8);
}

function formatPubDate(isoString: string): string {
  return new Date(isoString.replace(/-/g, "/")).toDateString();
}

interface IEnclosure {
  duration: number;
  link: string;
}

export interface IFeedItem {
  readonly guid: string;
  title: string;
  description: string;
  pubDate: string;
  enclosure: IEnclosure;
}

export interface IResultProps {
  result: Readonly<IFeedItem>;
  onClick: (feedItem: Readonly<IFeedItem>) => void;
}

export const Result: FunctionComponent<IResultProps> = (props: IResultProps) => {
  const { onClick, result } = props;
  const { description, title } = result;

  return (
    <li
      class={'result list-group-item list-group-item-action'}
      onClick={() => onClick(result)}
      tabIndex={0}>
      <a href={result.enclosure.link}>
        <h2 class='title' dangerouslySetInnerHTML={{ __html: title }} />
      </a>
      <strong class='pubDate'>{formatPubDate(result.pubDate)}</strong>
      <strong>&nbsp;&bull;&nbsp;</strong>
      <strong class='duration'>{formatDuration(result.enclosure.duration)}</strong>
      <p class='description' dangerouslySetInnerHTML={{ __html: description }} />
    </li>
  );
};
