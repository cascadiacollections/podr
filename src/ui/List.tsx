import { FunctionComponent, h } from "preact";
import { memo } from "preact/compat";
import { IFeedItem } from "./Result";
import { Result } from "./Result";

interface IListProps {
  results: ReadonlyArray<IFeedItem>;
  onClick: (item: Readonly<IFeedItem>) => void;
};

/**
 * List component that renders a table of podcast episodes
 * Using Pico's classless table approach for responsive design
 */
export const List: FunctionComponent<IListProps> = memo(
  (props: IListProps) => {
    const { results, onClick } = props;

    return (
      <section className="episodes-container">
        <table>
          <caption>Podcast Episodes</caption>
          <thead>
            <tr>
              <th scope="col">Episode</th>
              <th scope="col" className="date-column">Date</th>
              <th scope="col" className="duration-column">Duration</th>
            </tr>
          </thead>
          <tbody>
            {results.map((result: Readonly<IFeedItem>) => (
              <Result
                key={result.guid}
                result={result}
                onClick={onClick}
              />
            ))}
            {results.length === 0 && (
              <tr>
                <td colSpan={3}>No episodes available. Search or select a podcast to see episodes.</td>
              </tr>
            )}
          </tbody>
        </table>
      </section>
    );
  },
  // Custom comparison for memoization
  (prevProps, nextProps) => {
    // Only re-render if the results array has changed in length or content
    if (prevProps.results.length !== nextProps.results.length) {
      return false;
    }
    
    // Simple reference equality check is usually sufficient
    // since the results array is typically replaced entirely
    return prevProps.results === nextProps.results;
  }
);
