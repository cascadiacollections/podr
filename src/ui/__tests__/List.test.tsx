import { h } from 'preact';
import { render } from '@testing-library/preact';
import { List } from '../List';
import { IFeedItem } from '../Result';

// Mock the Result component to avoid rendering issues in tests
jest.mock('../Result', () => ({
  Result: ({ result, onClick }) => (
    <tr data-testid="mocked-result">
      <td>{result.title}</td>
      <td>{result.pubDate}</td>
      <td>{result.enclosure.duration}</td>
    </tr>
  ),
  // Re-export IFeedItem interface
  ...jest.requireActual('../Result')
}));

describe('List component', () => {
  const mockResults: IFeedItem[] = [
    {
      guid: 'test-guid-1',
      title: 'Test Episode 1',
      description: 'Description 1',
      pubDate: '2023-05-15T12:00:00Z',
      enclosure: {
        duration: 3600,
        link: 'https://example.com/episode1.mp3'
      }
    },
    {
      guid: 'test-guid-2',
      title: 'Test Episode 2',
      description: 'Description 2',
      pubDate: '2023-05-16T12:00:00Z',
      enclosure: {
        duration: 1800,
        link: 'https://example.com/episode2.mp3'
      }
    }
  ];

  test('renders with results', () => {
    const onClick = jest.fn();
    const { container } = render(<List results={mockResults} onClick={onClick} />);
    expect(container).toMatchSnapshot();
  });

  test('renders empty state', () => {
    const onClick = jest.fn();
    const { container } = render(<List results={[]} onClick={onClick} />);
    expect(container).toMatchSnapshot();
  });
});