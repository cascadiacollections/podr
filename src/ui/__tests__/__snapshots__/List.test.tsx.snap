// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders empty state 1`] = `
<div>
  <section
    class="episodes-container"
  >
    <table>
      <caption>
        Podcast Episodes
      </caption>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Episode
          </th>
          <th
            class="date-column"
            scope="col"
          >
            Date
          </th>
          <th
            class="duration-column"
            scope="col"
          >
            Duration
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            colspan="3"
          >
            No episodes available. Search or select a podcast to see episodes.
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;

exports[`List component renders with results 1`] = `
<div>
  <section
    class="episodes-container"
  >
    <table>
      <caption>
        Podcast Episodes
      </caption>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Episode
          </th>
          <th
            class="date-column"
            scope="col"
          >
            Date
          </th>
          <th
            class="duration-column"
            scope="col"
          >
            Duration
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          aria-label="Play episode: Test Episode 1"
          role="button"
          tabindex="0"
        >
          <td>
            <a
              aria-label="Stream or download: Test Episode 1"
              href="https://example.com/episode1.mp3"
            >
              Test Episode 1
            </a>
          </td>
          <td
            class="date-column"
          >
            <time
              datetime="2023-05-15T12:00:00Z"
            >
              Invalid Date
            </time>
          </td>
          <td
            class="duration-column"
          >
            01:00:00
          </td>
        </tr>
        <tr
          aria-label="Play episode: Test Episode 2"
          role="button"
          tabindex="0"
        >
          <td>
            <a
              aria-label="Stream or download: Test Episode 2"
              href="https://example.com/episode2.mp3"
            >
              Test Episode 2
            </a>
          </td>
          <td
            class="date-column"
          >
            <time
              datetime="2023-05-16T12:00:00Z"
            >
              Invalid Date
            </time>
          </td>
          <td
            class="duration-column"
          >
            00:30:00
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
